apps/aehttp/src/aehttp_cache_expires.erl

```diff
@@ -0,0 +1,218 @@
+aged_blocks_cache_time() ->
+    aeu_env:user_config_or_env([<<"http">>, <<"cache">>, <<"aged_blocks_cache_time">>],
+                               aehttp, [cache, aged_blocks_cache_time], ?DEFAULT_AGED_BLOCKS_CACHE_TIME).
+
+-spec handle_key_block_hash(cowboy_req:req(), any()) -> cowboy_expires_res().
+handle_key_block_hash(Req, State) ->
+    Hash0 = cowboy_req:binding(hash, Req),
+    case aeser_api_encoder:safe_decode(key_block_hash, Hash0) of
+        {ok, Hash} ->
+            Exp = expires_block_hash(Hash),
+            {Exp, Req, State};
+        _ -> {undefined, Req, State}
+    end.
+
+-spec handle_key_block_height(cowboy_req:req(), any()) -> cowboy_expires_res().
+handle_key_block_height(Req, State) ->
+    Height0 = cowboy_req:binding(height, Req),
+    case to_int(Height0) of
+        {ok, Height} ->
+            Exp = expires_block_height(Height),
+            {Exp, Req, State};
+        _ ->
+            {undefined, Req, State}
+    end.
+
+-spec handle_micro_block_hash(cowboy_req:req(), any()) -> cowboy_expires_res().
+handle_micro_block_hash(Req, State) ->
+    Hash0 = cowboy_req:binding(hash, Req),
+    case aeser_api_encoder:safe_decode(micro_block_hash, Hash0) of
+        {ok, Hash} ->
+            Exp = expires_block_hash(Hash),
+            {Exp, Req, State};
+        _ -> {undefined, Req, State}
+    end.
+
+-spec handle_transaction_hash(cowboy_req:req(), any()) -> cowboy_expires_res().
+handle_transaction_hash(Req, State) ->
+    Hash0 = cowboy_req:binding(hash, Req),
+    case aeser_api_encoder:safe_decode(tx_hash, Hash0) of
+        {ok, Hash} ->
+            case aec_chain:find_tx_with_location(Hash) of
+                {BlockHash, _Tx} when is_binary(BlockHash) ->
+                    Exp = expires_block_hash(BlockHash),
+                    {Exp, Req, State};
+                _ ->
+                    {undefined, Req, State}
+            end;
+        _ ->
+            {undefined, Req, State}
+    end.
+
+-spec expires_block_height(non_neg_integer()) -> {ok, aec_headers:header()} | undefined.
+expires_block_height(Height) ->
+    case aehttp_logic:get_key_header_by_height(Height) of
+        {ok, Header} -> expires_block_header(Header);
+        _ -> undefined
+    end.
+
+-spec expires_block_hash(binary()) -> {ok, aec_headers:header()} | undefined.
+expires_block_hash(Hash) ->
+    case aehttp_logic:get_header_by_hash(Hash) of
+        {ok, Header} -> expires_block_header(Header);
+        _ -> undefined
+    end.
+
+-spec expires_block_header(aec_headers:header()) -> non_neg_integer() | error.
+expires_block_header(Header) ->
+    case block_header_cache_time(Header) of
+        CacheTime when is_integer(CacheTime) ->
+            unixtime_to_datetime(CacheTime);
+        _ -> error
+    end.
+
+-spec block_header_cache_time(aec_headers:header()) -> non_neg_integer() | error.
+block_header_cache_time(Header) ->
+    Now = aeu_time:now_in_secs(),
+    BlockTimestamp = aec_headers:time_in_secs(Header),
```

Comment:
Potential problem here: we compare our local time with block's time synced from the network (which may have different clock time)

Reply:
That's a valid point. However we compare aggregated time period, which statistically given the protocol time constraints should be fine if `aged_blocks_time` is not set too low.